pkg/agent/cniserver/server.go

```diff
@@ -0,0 +1,386 @@
+	// parse request Stdin
+	conf := &types.NetConf{}
+	err := json.Unmarshal(request.Stdin, &conf)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// parse request Args
+	args := &CNIArgs{}
+	err = types.LoadArgs(request.Args, args)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return conf, args, err
+}
+
+func ParseResult(result *types100.Result) (*cnipb.CniResponse, error) {
+	var resultBytes bytes.Buffer
+	if err := result.PrintTo(&resultBytes); err != nil {
+		return nil, err
+	}
+	return &cnipb.CniResponse{
+		Result: resultBytes.Bytes(),
+		Error:  nil,
+	}, nil
+}
+
+func (s *CNIServer) CmdAdd(ctx context.Context, request *cnipb.CniRequest) (*cnipb.CniResponse, error) {
+	klog.Infof("Create new pod %s", request)
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	conf, args, err := ParseConf(request)
+	if err != nil {
+		klog.Errorf("Parse request conf error, err: %s", err)
+		return s.RetError(6, "Parse request conf error", err)
+	}
+
+	// require ipam for a new ip address
+	SetEnv(request)
+	r, err := ipam.ExecAdd("host-local", s.GetIpamConfByte(conf))
+	if err != nil {
+		klog.Errorf("could not allocate ip address, err: %s", err)
+		return s.RetError(7, "could not allocate ip address", err)
+	}
+	ipamResult, err := types100.NewResultFromResult(r)
+	if err != nil {
+		klog.Errorf("could not convert result, err: %s", err)
+		return s.RetError(6, "could not convert result", err)
+	}
+
+	// create cni result structure
+	result := &types100.Result{
+		CNIVersion: conf.CNIVersion,
+		IPs:        ipamResult.IPs,
+		Routes: []*types.Route{{
+			Dst: net.IPNet{
+				IP:   net.IPv4zero,
+				Mask: net.IPMask(net.IPv4zero)},
+			GW: ipamResult.IPs[0].Gateway}},
+		Interfaces: []*types100.Interface{{
+			Name:    request.Ifname,
+			Sandbox: request.Netns}},
+	}
+	// set the correspondence between interface and ip address
+	result.IPs[0].Interface = types100.Int(0)
+
+	nsPath := "/host" + request.Netns
+	// vethName - ovs port name
+	vethName := "veth" + strings.Split(request.Netns, "/")[2]
+	if err = ns.WithNetNSPath(nsPath, func(hostNS ns.NetNS) error {
+		// create veth pair in container NS and host NS
+		// TODO: MTU is a const variable here
+		_, containerVeth, err := ip.SetupVethWithName(request.Ifname, vethName, 1500, "", hostNS)
+		if err != nil {
+			klog.Errorf("create veth device error, err: %s", err)
+			return err
+		}
+		result.Interfaces[0].Mac = containerVeth.HardwareAddr.String()
+		if err = ipam.ConfigureIface(request.Ifname, result); err != nil {
+			klog.Errorf("configure ip address in container error, err: %s", err)
+			return err
+		}
+		return nil
+	}); err != nil {
+		return s.RetError(1, "exec error in namespace", err)
+	}
+
+	// add the veth device to ovs bridge
+	err = s.ovsDriver.CreatePort(vethName, "", 0)
+	if err != nil {
+		klog.Errorf("create ovs port error, vethName: %s, err: %s", vethName, err)
+		return s.RetError(5, "exec error in namespace", err)
```

Comment:
Could we use the ErrorCode instead of number?

Reply:
okay, done